### PR TITLE
Support synthetic "time-in-queue" for RabbitMQ

### PR DESCRIPTION
**Pull request by bantonsson**
It's now possible to get the _real_ service name on the `amqp.command` spans, as well as a `amqp.deliver` span with service name `rabbitmq` that represents the time spent in the queue, by setting `-Ddd.rabbit.legacy.tracing.enabled=false`, or `DD_RABBIT_LEGACY_TRACING_ENABLED=false`.

Please note that enabling the new RabbitMQ tracing behavior will most likely change the [Service Entry Span](https://docs.datadoghq.com/tracing/visualization/#service-entry-span) which affects traces are automatically kept by the [Intelligent Retention Filter](https://docs.datadoghq.com/tracing/trace_retention_and_ingestion/#datadog-intelligent-retention-filter).

### CODE REVIEW

File: dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java [hunk 1]
```diff
@@ -1,38 +1,58 @@
...
 import datadog.trace.bootstrap.instrumentation.api.Tags;
 import datadog.trace.bootstrap.instrumentation.api.UTF8BytesString;
 import datadog.trace.bootstrap.instrumentation.decorator.ClientDecorator;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
 
 public class RabbitDecorator extends ClientDecorator {
 
   public static final CharSequence AMQP_COMMAND = UTF8BytesString.create("amqp.command");
+  public static final CharSequence AMQP_DELIVER = UTF8BytesString.create("amqp.deliver");
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java:31]: Is this a reasonable name for the span spent in the queue?

File: dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java [hunk 2]
```diff
@@ -100,15 +117,114 @@ public void onDeliver(final AgentSpan span, final String queue, final Envelope e
...
+      span = startSpan(AMQP_COMMAND, spanStartMicros);
+    }
+    if (null != body) {
+      span.setTag("message.size", body.length);
+    }
+    // TODO - do we still need both?
+    if (null != properties && null != properties.getTimestamp()) {
+      // this will be set if the sender sets the timestamp,
+      // or if a plugin is installed on the rabbitmq broker
+      long produceMillis = properties.getTimestamp().getTime();
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java:197]: We could use the `queueStartMillis` here if there was no `timeStamp` available.

File: dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java (hunk 1 above)
Comment (mcculls) [reply dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java:35]: I've used the same name (without the `amqp.` prefix) in the JMS "time-in-queue" PR, and I think it's a reasonable name for the "time-in-queue" span

File: dd-trace-core/src/main/java/datadog/trace/common/writer/TraceStructureWriter.java [hunk 3]
```diff
@@ -34,16 +36,41 @@ public TraceStructureWriter(String outputFile) {
...
+      String[] args = ARGS_DELIMITER.split(outputFile);
+      String fileName = args[0];
       this.out =
-          outputFile.isEmpty() || outputFile.equals(":")
+          fileName.isEmpty()
               ? System.err
-              : new PrintStream(
-                  new FileOutputStream(new File(Strings.replace(outputFile, ":", ""))));
+              : new PrintStream(new FileOutputStream(new File(fileName)));
+      if (args.length > 1) {
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/common/writer/TraceStructureWriter.java:54]: this `if` could be dropped as it's the same as the `for` loop boundary check when `i == 1`

Comment (bantonsson) [reply dd-trace-core/src/main/java/datadog/trace/common/writer/TraceStructureWriter.java:54]: 🤦 Thanks

File: dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/test/groovy/RabbitMQTest.groovy [hunk 4]
```diff
@@ -338,129 +375,172 @@ class RabbitMQTest extends AgentTestRunner {
...
+      case "deliver":
+        def consumer = new StringConsumer(channel)
+        channel.basicConsume(queueName, consumer)
+        body = consumer.body
+        expectedTraces = 3
+        break
+      default:
+        break
+    }
+    //expectedTraces += hasQueueSpan() && !noParent ? 1 : 0
```
Comment (mcculls) [inline dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/test/groovy/RabbitMQTest.groovy:424]: safe to remove?

File: dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java [hunk 5]
```diff
@@ -100,15 +121,124 @@ public void onDeliver(final AgentSpan span, final String queue, final Envelope e
...
+    final AgentSpan span;
+    if (null != parentContext) {
+      span = startSpan(AMQP_COMMAND, parentContext, spanStartMicros);
+    } else {
+      span = startSpan(AMQP_COMMAND, spanStartMicros);
+    }
+    if (null != body) {
+      span.setTag("message.size", body.length);
+    }
+    // TODO - do we still need both?
```
Comment (mcculls) [inline dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java:201]: is this in relation to the `RECORD_QUEUE_TIME_MS` tag? - I've left that in for JMS for now. Once the new "time-in-queue" structure is the default we can think about deprecating that tag and eventually remove it.

**Review by mcculls — APPROVED** [review]
Thanks for the smoke-test improvements!

File: dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java (hunk 5 above)
Comment (bantonsson) [reply dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/main/java/datadog/trace/instrumentation/rabbitmq/amqp/RabbitDecorator.java:201]: That is an old comment wrt to the `RECORD_QUEUE_TIME_MS` and the `RECORD_END_TO_END_DURATION_MS`, that I'm not completely sure what it refers to. I agree that we can flip this around and eventually remove `RECORD_QUEUE_TIME_MS` when _"time-in-queue"_ is the default.

**Review by charliegracie — APPROVED** [review]

**Pull request merged**